### ISSUE TITLE
feat: gzip s3

### ISSUE THREAD

**Issue by sja** [danielcondemarin/serverless-nextjs-plugin#28]
Hi!
Would a PR be merged which compresses some assets by mime type (*/text, css, svg) before uploading to s3? The upload config then has two options: The `encoding`- and the `content-type`-header. Thats all. To configure this, I suggest a flag in `serverless.yaml` with a bool to enable this (what should be the default?) and a list to override the mime types.

What do you think?

**Comment by danielcondemarin** [comment]
Would it not be better to add CloudFront support? It already has compression etc. plus another benefits like custom domains for the assets bucket etc.

**Comment by sja** [comment]
You're right.

**Comment by victorhqc** [comment]
> Would it not be better to add CloudFront support? It already has compression etc. plus another benefits like custom domains for the assets bucket etc.

I'm having some issues with Cloud-front as it caches the headers sent from the client as well. This is important in my website as I need those to either render the mobile or desktop website. By caching those results the website incorrectly renders a website depending on the cached headers, not the real ones being sent.

I know that this could be solved by having a different domain for mobile, like `https://m.mywebiste.com`. However, having an option to upload gzip to S3 and not using Cloud-front could be a different option to solve a problem.

What do you think?

**Comment by mattdell** [comment]
> You're right.

Is there an issue to track the progress of adding cloudfront support then?

I'm happy with just S3 gzip support. Simple and effective.

**Comment by danielcondemarin** [comment]
@mattdell It might take a while before I can take a look at supporting CloudFront. 

@sja / others interested. I'm happy to accept a PR for this, but first let's get an example serverless.yml with the new options.

**Comment by mattdell** [comment]
Is there anything I can do to help with this? Push me in the right direction and I may be able to help as I desperately need compression for my site.

**Comment by danielcondemarin** [comment]
@mattdell I would suggest trying a solution using https://github.com/webpack-contrib/compression-webpack-plugin. Then we can look at a plugin option to set the `Content-Encoding` header on the compressed assets.

**Comment by mattdell** [comment]
For anyone wanting gzip look at PR #92 which introduces feature parity with NextJS and allows for gzip if following the `static` and `public` convention. 

You'll just need to remove `assetPrefix` if currently using it.

**Comment by danielcondemarin** [comment]
@mattdell How does it allow for gzip 🤔 ? The resources in public/ and static/ are not gzipped when uploaded to S3 ... 

The current options I see for gzipping are:
- Upload gzipped static files to S3
or 
- Support CloudFront and let it do the gzipping

**Comment by mattdell** [comment]
Assets served through API Gateway are gzipped by default, so if you're not using the `assetPrefix` and using a relative path then content should get encoded without needing to do anything.

**Comment by danielcondemarin** [comment]
> Assets served through API Gateway are gzipped by default, so if you're not using the `assetPrefix` and using a relative path then content should get encoded without needing to do anything.

Cool! Although seems like you still have to define `minimumCompressionSize` according to https://serverless.com/framework/docs/providers/aws/events/apigateway/#api-gateway

**Comment by mattdell** [comment]
Yes. That is a good point. I do have that set.

```
provider:
  name: aws
  runtime: nodejs8.10
  stage: ${env:STAGE}
  region: eu-west-1
  apiGateway:
    minimumCompressionSize: 1024
```

**Comment by sja** [comment]
What I did not understood while creating this Issue: Just uploading the assets gzipped is not a good solution. It does no header-negotiation which detects that the requesting browser supports unzipping it. Additionally, there are more than one algorithms in modern browsers for compression, gzip does not have to be the best one.
For me, CloudFront does all these things together with caching the assets on edge from my s3 bucket. The API and HTML requests which are delivered by the lambdas are compressed by the mentioned API gateway settings.

**Comment by danielcondemarin** [comment]
Closing this as I've just published v2.1.0-alpha.0 with cloudfront support and compression enabled by default. Documented [here](https://github.com/danielcondemarin/serverless-nextjs-plugin#serving-static-assets).

Please try the new cloudfront support and feedback! Thanks 🙏
